docs/notebooks/recovering_from_errors.pct.py

```diff
@@ -0,0 +1,144 @@
+# %% [markdown]
+# # Recovering from errors during optimization
+
+# %%
+import numpy as np
+import tensorflow as tf
+import random
+
+np.random.seed(1793)
+tf.random.set_seed(1793)
+random.seed(0)
+
+# %% [markdown]
+# Sometimes the Bayesian optimization process encounters an error from which we can recover, without the need to restart the run from the beginning. In this tutorial, we'll simulate such an error and show how to recover from it.
+#
+# We'll use a similar setup to the [EI notebook](expected_improvement.ipynb), but use an observer that intermittently breaks when evaluated, and needs manual attention to get running again. We can simulate fixing the observer with its `manual_fix` method.
+
+# %%
+import trieste
+from trieste.acquisition.rule import OBJECTIVE
+from trieste.utils.objectives import branin
+
+
+class FaultyBranin:
+    def __init__(self):
+        self._is_broken = False
+
+    def manual_fix(self):
+        self._is_broken = False
+
+    def __call__(self, x):
+        if self._is_broken:
+            raise Exception("Observer is broken")
+
+        if random.random() < 0.3:
+            self._is_broken = True
+
+        return {OBJECTIVE: trieste.data.Dataset(x, branin(x))}
+
+
+observer = FaultyBranin()
+
+# %% [markdown]
+# ## Set up the problem
+# We'll use the same set up as before, except for the acquisition rule, where we'll use `TrustRegion`, which will better illustrate how to recover.
+
+# %%
+import gpflow
+
+search_space = trieste.space.Box(
+    tf.cast([0.0, 0.0], tf.float64), tf.cast([1.0, 1.0], tf.float64)
+)
+initial_data = observer(search_space.sample(5))
+
+variance = tf.math.reduce_variance(initial_data[OBJECTIVE].observations)
+kernel = gpflow.kernels.Matern52(variance, [0.2, 0.2]) + gpflow.kernels.White(1e-12)
+gpr = gpflow.models.GPR(
+    initial_data[OBJECTIVE].astuple(), kernel, noise_variance=1e-5
+)
+gpflow.set_trainable(gpr.likelihood, False)
+models = {OBJECTIVE: trieste.models.GaussianProcessRegression(gpr)}
+
+acquisition_rule = trieste.acquisition.rule.TrustRegion()
+
+# %% [markdown]
+# ## Run the optimization loop
+#
+# The optimizer returns an `OptimizationResult`. We can use that object's `astuple` method to get a 2-tuple of the final result and the history of the successful optimization steps.
+
+# %%
+bo = trieste.bayesian_optimizer.BayesianOptimizer(observer, search_space)
+
+result, history = bo.optimize(15, initial_data, models, acquisition_rule).astuple()
+
+# %% [markdown]
+# We can see from the logs that the optimization loop failed, and this can be sufficient to know what to do next if we're working in a notebook. However, sometimes our setup means we don't have access to the logs. We'll pretend from here that's the case.
```

Comment:
How can the user see that from the logs ?

Reply:
It will appear when run as a notebook. I've written this primarily for notebook users/people reading the built docs

Reply:
ok, up to you whether you think that's obvious enough by itself or whether it might be helpful to clarify that "logs" refers to "what bo.optimize() prints to stdout [rather than some logfile that I have to dig up somewhere]"?